MAINT: remove error throw in binned_statistic_dd() on non-finite values or samples

rlucas7 opened the pull request:
#### Reference issue
Closes gh-11365

#### What does this implement/fix?
When I added an error throw in gh-10664 it wasn't the right fix and broke the the `lightkurve` package, see e.g.
https://github.com/KeplerGO/lightkurve/pull/661
and 
https://github.com/KeplerGO/lightkurve/issues/663

The proposal from @barentsen as a fix is confirmed here and the test for non-finite values is repurposed to check and raise a `ValueError` if the `bins` arg is `int` type and there are any non-finite `values`. 

#### Additional information
The error is raised in `binned_statistic_dd()` so it is straightforward to the user but the underlying issue is that `_bin_edges()` is ***not*** robust to a specified number of bins and ***any*** non-finite values. A follow up PR may want to provide a more permanent fix to `_bin_edges()`-if possible.

tylerjereddy commented:
Cool, maybe another `stats` person will notice now that the change is proposed in a PR. Otherwise, I'll try to take a closer look. Sounds like you may have reached agreement with the issue submitter in any case.

rlucas7 commented:
I’ll send an email notifying on the dev list hopefully that will get some more eyes. 

> On Jan 16, 2020, at 11:27 PM, Tyler Reddy <notifications@github.com> wrote:
> 
> ﻿
> Cool, maybe another stats person will notice now that the change is proposed in a PR. Otherwise, I'll try to take a closer look. Sounds like you may have reached agreement with the issue submitter in any case.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rlucas7 commented:
Removing the "needs decision" tag, there seems to be general agreement this was an unintential defect that I introduced. Agreement is to backport in 1.4.2 and include in 1.5.0 release too. 
See e.g. discussion in associated issue.

WarrenWeckesser commented:
@rlucas7, could you incorporate the changes that I suggested in https://github.com/scipy/scipy/pull/10664 into this PR?  The function should not reference `values` when `statistic` is `'count'`.

Also, double-check that the code you add complies with PEP-8 (max. line length 79, spaces after commas, etc.).  I often temporarily rename `tox.ini` to something like `tox.ini.save`, and then run `pycodestyle` on the files that I'm editing to ensure that I don't introduce any new PEP-8 violations.

rlucas7 commented:
> _EDIT_
> @rlucas7, could you incorporate the changes that I suggested in #10664 into this PR? The function should not reference `values` when `statistic` is `'count'`. Sorry, I see that the pull request already fixes that, by simply not checking `values` for finiteness. I'll look into [gh-11365](https://github.com/scipy/scipy/issues/11365) before commenting further.
> 
> Also, double-check that the code you add complies with PEP-8 (max. line length 79, spaces after commas, etc.). I often temporarily rename `tox.ini` to something like `tox.ini.save`, and then run `pycodestyle` on the files that I'm editing to ensure that I don't introduce any new PEP-8 violations.

I'll check the PEP 8 and update the 2d binned statistic docstring example you found on the original PR when I get home.

rlucas7 commented:
@WarrenWeckesser I reverted the docstring changes in the `binned_statistic_2d` function, the second instance of `x` in the function call is now reverted to `None` in both instances. I also made the PEP 8 changes. There where E501 issues in the doctest results, the `BinnedStatisticResult` return values were all on 1-line. I've continued them across 2 (or 3 if necessary) lines. Hopefully this doesn't make the doctests fail in the build, I tested them locally after the chnages via `python  scipy/stats/_binned_statistic.py ` which gave no output so hopefully these are fine. 

Also, thanks for the tip on the tox file. Indeed that has a setting to ignore the E501 in pycodestyle setting. I didn't realize there were pycodestyle ignore flags inside that file-good to know.

rlucas7 commented:
@WarrenWeckesser I've addressed the PEP8 issues in doctests and those are not failing on latest build. The 2 travis failures appear to be related to a curvefit covariance test failing, nothing that I've changed. Would be great if you could take another look at the PR when you have time.

rlucas7 commented:
The previous test errors were unrelated (special and linalg tests) and the branch was about 10 days old so I rebased.

rlucas7 commented:
looks like the tests all pass in travis on rebase. 

Only windows failure is:
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=5316&view=logs&j=689af343-93ec-58dc-880e-8a2e7b5f2ce5&t=0daec287-0840-5bd9-6022-6f148a04400b&l=478
and this PR does not touch the differential evolution solver.

Leaving as is until @WarrenWeckesser has time to review the changes he suggested.

WarrenWeckesser commented:
@rlucasy, finally got back to this.  Looks good.

The master branch now has a couple lines added to the unit test to avoid making the inplace change of `self.u` visible after the test is finished.  That could be simplified by using `u = self.u.copy()`, and then use the code in this PR for the rest of the test.  Or do the swapping that is in master--either is fine.

WarrenWeckesser commented on scipy/stats/_binned_statistic.py:
```diff
@@ -513,8 +517,9 @@ def binned_statistic_dd(sample, values, statistic='mean',
-    if not np.isfinite(values).all() or not np.isfinite(sample).all:
-        raise ValueError('%r or %r contains non-finite values.' % (sample, values,))
+    # NOTE: for _bin_edges(), see e.g. gh-11365
+    if isinstance(bins, int) and not np.isfinite(sample).all():
```
Taking another look, I just noticed this: `isinstance(bins, int)` will not recognize numpy integers, e.g. `np.int64(10)`.  Lately we have settled on checking for integerness by passing the value through `operator.index`.   That is, if `operator.index(bins)` doesn't raise a `TypeError`, then `bins` is an integer.  So at a minimum, we could add this before the current line:
```
    try:
        bins = operator.index(bins)
    except TypeError:
        # bins is not an integer
        pass
    # If bins was an integer-like object, now it is an actual Python int.
```
Then the test `isinstance(bins, int)` will work.  There is probably a prettier way to do that, but it would work.

rlucas7 replied:
> @rlucasy, finally got back to this. Looks good.
> 
> The master branch now has a couple lines added to the unit test to avoid making the inplace change of `self.u` visible after the test is finished. That could be simplified by using `u = self.u.copy()`, and then use the code in this PR for the rest of the test. Or do the swapping that is in master--either is fine.

I finally got a few spare moments to look at this PR again, sorry it's taken me awhile. I rebased and resolved the conflicts as you suggest. However, the test you are referencing should be removed as it was incorrect to throw the error to begin with. I'll do a force push today with the resolution of this and the other issue you noted.

rlucas7 replied:
> Taking another look, I just noticed this: `isinstance(bins, int)` will not recognize numpy integers, e.g. `np.int64(10)`. Lately we have settled on checking for integerness by passing the value through `operator.index`. That is, if `operator.index(bins)` doesn't raise a `TypeError`, then `bins` is an integer. So at a minimum, we could add this before the current line:
> 
> ```
>     try:
>         bins = operator.index(bins)
>     except TypeError:
>         # bins is not an integer
>         pass
>     # If bins was an integer-like object, now it is an actual Python int.
> ```
> 
> Then the test `isinstance(bins, int)` will work. There is probably a prettier way to do that, but it would work.

Thanks for catching this Warren, always something with the types it seems :) 
I'll add a test for this and the suggested fix you've noted here. There maybe a prettier way to do it but you're in good company with your suggestion. see e.g.: https://stackoverflow.com/a/54381905/3164100 so I see no reason to try something fancy.

rlucas7 commented:
@WarrenWeckesser I added the changes you've suggested. Thanks for catching that one. 
I only added a test case within the existing non-finite test and only for `np.int64`. 
I'm guessing `np.int` is a wrapper for python's `int` because that was the only one I tested that seemed to pass the check. Although, `np.int8`, `np.int32` etc. all would've failed I figured testing only one of the int types from numpy should be sufficient. If you want me to add the others in I can-just seems like overkill.

WarrenWeckesser commented:
> I'm guessing np.int is a wrapper for python's int

`np.int` *is* the Python `int` object, and that has occasionally been a source of confusion over the years.  This aliasing of Python's builtin types in the `numpy` namespace may soon be deprecated: https://github.com/numpy/numpy/pull/14882

> Although, np.int8, np.int32 etc. all would've failed I figured testing only one of the int types from numpy should be sufficient.

Yup, that's fine.

WarrenWeckesser commented on scipy/stats/tests/test_binned_statistic.py:
```diff
@@ -49,16 +49,20 @@ def test_big_number_std(self):
-        assert_raises(ValueError, binned_statistic, x, u, 'std', bins=10)
+        assert_raises(ValueError, binned_statistic, u, x, 'std', bins=10)
+        # need to test for non-python specific ints, e.g. np.int8, np.int64
+        # note np.int will pass the isinstance(bins, int) test but other int no
```
```suggestion
```

WarrenWeckesser commented on scipy/stats/tests/test_binned_statistic.py:
```diff
@@ -49,16 +49,19 @@ def test_big_number_std(self):
         u[0] = np.nan
-        assert_raises(ValueError, binned_statistic, x, u, 'count', bins=10)
+        assert_raises(ValueError, binned_statistic, u, x, 'count', bins=10)
+        # needs to raise on non language int types too, e.g np.int
```
```suggestion
```

WarrenWeckesser commented on scipy/stats/tests/test_binned_statistic.py:
```diff
@@ -49,16 +49,18 @@ def test_big_number_std(self):
-        assert_raises(ValueError, binned_statistic, x, u, 'std', bins=10)
+        assert_raises(ValueError, binned_statistic, u, x, 'std', bins=10)
+        # need to test for non-python specific ints, e.g. np.int8, np.int64
+        assert_raises(ValueError, binned_statistic, u, x, 'std', bins=np.int64(10))
```
```suggestion
        assert_raises(ValueError, binned_statistic, u, x, 'std',
                      bins=np.int64(10))
```

WarrenWeckesser commented:
@rlucas7, thanks! I removed a couple comments that weren't really needed, fixed a line length, squashed and merged.

rlucas7 commented:
Thanks Warren, LGTM.

Sincerely, 

-Lucas Roberts 

> On May 11, 2020, at 12:53 AM, Warren Weckesser <notifications@github.com> wrote:
> 
> ﻿
> @rlucas7, thanks! I removed a couple comments that weren't really needed, fixed a line length, squashed and merged.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
